lib/LanguageServerReferenceFinder/Handler/ReferencesHandler.php

```diff
@@ -157,7 +161,7 @@ public function references(
 
     public function registerCapabiltiies(ServerCapabilities $capabilities): void
     {
-        $capabilities->referencesProvider = true;
+        $capabilities->referencesProvider = null !== $this->clientCapabilities->textDocument->references;
```

Comment:
is this correct? the server is still capable, the client is not?

Reply:
I did not found any better solution to prevent the client from calling to phpactor given method.

Reply:
ah I see.

the goal here is f.e. disable completion and use another LS for that?

I don't think this is about client _capabilities_ (I don't think that can necessarily be controlled by the user in every IDE?). I think this would then need to be a Phpactor level config, or mechanism to enable / disable extensions/handlers.

Reply:
I still work on a legacy project mentioned in https://github.com/phpactor/phpactor/issues/1125 Calling for completion or references still takes unacceptable time (with index on in-memory FS, JIT and  `ComposerAutoloaderExtension`) and warms up CPU.

This forces me to use intelephense instead for these costly requests. 

This change help me to use phpactor anywhere on that project (and then possibly learn it, find bugs etc.).

Reply:
Maybe not every IDE can this but LSP seems to allow it:

> ClientCapabilities define capabilities for dynamic registration, workspace and text document features the client supports. The experimental can be used to pass experimental capabilities under development. For future compatibility a ClientCapabilities object literal can have more properties set than currently defined. Servers receiving a ClientCapabilities object literal with unknown properties should ignore these properties. A missing property should be interpreted as an absence of the capability. If a missing property normally defines sub properties, all missing sub properties should be interpreted as an absence of the corresponding capability.

Reply:
> text document features the client supports

but the client defines what the client supports not the user? in your case _you_ want to disable (f.e.) text completion. the client still supports it.

in this PR the client determines what the server supports, which seems wrong.

Reply:
(i definitely support enabling/disabling functionality btw, there is a good use case for disabling completion and using another LS for that)

Reply:
> the client still supports it.

I understand the word "client" as a concrete instance (so it supports at most those capabilities which it declared at initialization) - not as a software able to create such (variously configured) instances. LSP-compliant server seems to know only about the declared set. 

>  A missing property should be interpreted as an absence of the capability.

> in this PR the client determines what the server supports

Yes, it (at this PR finish) would restrict the set according to the rule quoted above.

Reply:
```
$capabilities->referencesProvider = null !== $this->clientCapabilities->textDocument->references;
```

read literally, this says "I support references only if the client supports it". But that's a lie, it supports references regardless of if the client supports it.

As far as I understand it (and I can't see any contraditction in that spec extract). Capabilties braodcast what the server/client support. They do not intefere with eachother. We use the client capabilities to see if the client supports certain messages from the server. If the client doesn't support certain features, the server won't send those messages.

Unless I'm wrong I would rather we added explicit Phpactor configuration to enable/disable handlers somehow, as this would implicitly inform the ServerCapabilties (because they simply would not be registered, and the RPC methods would not be exposed) and the Client would then not use the server for f.e. text completion.

Reply:
>  If the client doesn't support certain features, the server won't send those messages.
It's what I try to implement there. For example, client (understood as a configured instance) does not support `textDocument/references` and then the server in the response says that it does not provide it to the client on its demand. Maybe, technically, in this PR the server can still response for `textDocument/references` by some tricky way (is it mentioned lie?). Maybe we need to prevent from execution this method in such ways. 

What else is the purpose of sending client LSP-defined capabilities than making an intersection with the server-provided set in the result? 

See also
https://www.reddit.com/r/neovim/comments/nvds8f/disable_single_capabilities_for_a_given_lsp_server/
https://github.com/microsoft/language-server-protocol/issues/940

I still do not understand if `dynamicRegistration` can be related to this issue:
https://medium.com/@malintha1996/understanding-the-language-server-protocol-5c0ba3ac83d2#6c58